modules/fedimint-mint-client/src/lib.rs

```diff
@@ -843,6 +851,70 @@ impl MintClientModule {
         stream.next_or_pending().await
     }
 
+    /// Provisional implementation of note consolidation
+    ///
+    /// When a certain denomination crosses the threshold of notes allowed,
+    /// spend some chunk of them as inputs.
+    pub async fn consolidate_notes(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        operation_id: OperationId,
+    ) -> anyhow::Result<Vec<ClientInput<MintInput, MintClientStateMachines>>> {
+        let notes_by_denomination: BTreeMap<Amount, Vec<SpendableNote>> = dbtx
+            .find_by_prefix_sorted_descending(&NoteKeyPrefix)
+            .await
+            .map(|(key, note)| (key.amount, note))
+            .fold(
+                BTreeMap::<Amount, Vec<SpendableNote>>::new(),
+                |mut acc, (amount, note)| async move {
+                    acc.entry(amount).or_default().push(note);
+                    acc
+                },
+            )
+            .await;
+
+        /// At how many notes of the same denomination should we try to
+        /// consolidate
+        const MAX_NOTES_PER_TIER: usize = 10;
+        /// Number of notes per tier to leave after threshold was crossed
+        const MIN_NOTES_PER_TIER: usize = 4;
```

Comment:
The reasoning for this hysteresis would be interesting.

Imo it makes sense to use a power (of the base used for tier generations) difference between these values, so all the notes get consolidated into a higher tier. But the delta is 6 here and we should also not hard-code the assumption that we deal with base 2 denominations imo.

Reply:
I was hopping to minimize churn here. We definitely don't want to just take one note from a tier that happened to go over the limit, because it might just get remitted as a same tier.

Or if tier X gets consolidated to X+1, then in next TX tier X+1 to X+2, etc. that's a lot of churn, while it would be better to go straight to X+4 or higher.

Though I now think I should first see if any tier crossed HIGH, and then take all notes above LOW from *any* tier, up to LIMIT of total notes. Basically two passes: check for overflow, start grabbing notes.

I didn't do it the first time around, because I didn't have convenient APIs, which I added later down the line.
@elsirion